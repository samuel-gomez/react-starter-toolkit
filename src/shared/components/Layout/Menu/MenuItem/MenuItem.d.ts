/* eslint-disable @typescript-eslint/no-explicit-any */
import { createMenu } from './MenuItem.container';

export type TActionElt = {
  url?: string;
  label: string;
};

// le container reçoit les items simples
export type TMenuItemBase = TActionElt & {
  children?: TMenuItemBase[];
  classModifier?: string;
  createMenuFn?: typeof createMenu;
  basePathChildren?: string;
  basePath?: string;
  classModifierItem?: string;
};

// la vue reçoit les items avec des props fournies par le container
export type TMenuItem = TMenuItemBase & {
  ariaHaspopup?: any;
  ariaExpanded?: any;
  ariaLabel?: string;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
};

export type TMenuItemEnhanced = TMenuItemBase & {
  authorized?: string[];
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
};
