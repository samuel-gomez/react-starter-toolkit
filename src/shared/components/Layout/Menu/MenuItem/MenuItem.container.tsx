import { isEmpty } from 'lodash';
import Authorize from 'shared/components/Authorize';
import { MenuItem } from './MenuItem';
import type { TMenuItemEnhanced, TMenuItemBase } from './MenuItem.d';

export const createMenu = (menuItems: TMenuItemBase[], onClick?: (event: React.MouseEvent<HTMLElement>) => void, basePath?: string) =>
  menuItems.map(menuItem => (
    <MenuItemEnhanced key={menuItem.label} basePath={basePath} {...menuItem} onClick={onClick} classModifierItem={menuItem.classModifierItem} />
  ));

const MenuItemEnhanced = ({
  authorized,
  url,
  label,
  classModifier,
  children,
  basePath,
  classModifierItem,
  createMenuFn = createMenu,
  ...props
}: TMenuItemEnhanced) => {
  const newClassModifier = [classModifier, classModifierItem, children ? 'haschild' : ''].join(' ').trim();
  const ariaProps = url ? { ariaHaspopup: 'true', ariaExpanded: 'false', ariaLabel: label } : {};
  const newUrl = !isEmpty(basePath) && !isEmpty(url) ? `${basePath}/${url}` : url;

  return (
    <Authorize authorized={authorized}>
      <MenuItem {...props} basePath={basePath} createMenuFn={createMenuFn} url={newUrl} label={label} classModifier={newClassModifier} {...ariaProps}>
        {children}
      </MenuItem>
    </Authorize>
  );
};

export default MenuItemEnhanced;
