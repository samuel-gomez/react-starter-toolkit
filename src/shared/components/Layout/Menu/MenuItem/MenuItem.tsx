import Link from 'next/link';
import { NavBarItem } from '@axa-fr/react-toolkit-layout-header';
import type { TActionElt, TMenuItem } from './MenuItem.d';

export const ActionElt = ({ url, label }: TActionElt) =>
  url ? (
    <Link className="af-nav__link" href={url}>
      {label}
    </Link>
  ) : (
    <span className="af-nav__link">{label}</span>
  );

export const MenuItem = ({ children, basePath, url, label, ariaHaspopup, ariaExpanded, ariaLabel, createMenuFn, onClick, ...rest }: TMenuItem) => (
  <NavBarItem
    {...rest}
    actionElt={ActionElt({ url, label })}
    onClick={onClick}
    aria-haspopup={ariaHaspopup}
    aria-expanded={ariaExpanded}
    aria-label={ariaLabel}
  >
    {children && createMenuFn?.(children, onClick, basePath)}
  </NavBarItem>
);

export default MenuItem;
