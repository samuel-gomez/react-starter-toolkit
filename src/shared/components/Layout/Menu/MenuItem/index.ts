export { default, createMenu } from './MenuItem.container';
export { MenuItem } from './MenuItem';
export type { TActionElt, TMenuItem, TMenuItemEnhanced, TMenuItemBase } from './MenuItem.d';
