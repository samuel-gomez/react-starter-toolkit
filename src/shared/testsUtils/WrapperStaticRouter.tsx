/* eslint-disable react/display-name */
import { ReactNode } from 'react';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';

const WrapperStaticRouter =
  (route = '/') =>
  ({ children }: { children: ReactNode }) =>
    <MemoryRouterProvider url={route}>{children}</MemoryRouterProvider>;

export default WrapperStaticRouter;
