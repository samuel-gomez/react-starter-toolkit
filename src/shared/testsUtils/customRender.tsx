/* eslint-disable react/display-name */
import { render, RenderOptions } from '@testing-library/react';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';

import FetchProvider from 'providers/FetchProvider';
import UserProvider from 'providers/UserProvider';
import { ReactElement, ReactNode } from 'react';
import { MOCK_API_URL } from '.';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string;
};

const MockProviders =
  ({ role = '', ...testMock }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const useOidcAccessTokenFn = jest.fn().mockReturnValue({ accessToken: 'accessToken' });
    const useOidcUserFn = jest.fn().mockReturnValueOnce({ oidcUser: { member_of: [`CN=${role}`], name: 'Samuel Gomez' } });
    return (
      <UserProvider useOidcUserFn={useOidcUserFn}>
        <FetchProvider
          apiUrl={MOCK_API_URL}
          fetchConfig={{
            headers: { testMock: JSON.stringify(testMock) },
          }}
          useOidcAccessTokenFn={useOidcAccessTokenFn}
        >
          <MemoryRouterProvider url="">{children}</MemoryRouterProvider>
        </FetchProvider>
      </UserProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// override render method
export { customRender as render };
