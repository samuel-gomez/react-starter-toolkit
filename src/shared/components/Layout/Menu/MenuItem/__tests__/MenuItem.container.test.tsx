import { act } from '@testing-library/react';
import { renderWithWrapperStaticRouter } from 'shared/testsUtils';
import MenuItemEnhanced from '../MenuItem.container';
import { menuItemsMock } from './MenuItem.mock';

describe('<MenuItemEnhanced/>', () => {
  it('Render <Menu/>', async () => {
    const { asFragment } = renderWithWrapperStaticRouter(<MenuItemEnhanced label="test" />);
    act(() => expect(asFragment()).toMatchSnapshot());
  });

  it('Render <Menu/> hasChild', async () => {
    const { asFragment } = renderWithWrapperStaticRouter(<MenuItemEnhanced label="test">{menuItemsMock}</MenuItemEnhanced>);
    act(() => expect(asFragment()).toMatchSnapshot());
  });

  it('Render <Menu/> with', async () => {
    const { asFragment } = renderWithWrapperStaticRouter(<MenuItemEnhanced label="test" basePath="/demos" url="path" />);
    act(() => expect(asFragment()).toMatchSnapshot());
  });
});
