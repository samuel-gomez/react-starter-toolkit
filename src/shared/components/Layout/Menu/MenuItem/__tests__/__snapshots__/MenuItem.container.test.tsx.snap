// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItemEnhanced/> Render <Menu/> 1`] = `
<DocumentFragment>
  <li
    class="af-nav__item "
    role="none"
  >
    <span
      aria-expanded="false"
      class="af-nav__link"
      role="menuitem"
      tabindex="-1"
    >
      test
    </span>
  </li>
</DocumentFragment>
`;

exports[`<MenuItemEnhanced/> Render <Menu/> hasChild 1`] = `
<DocumentFragment>
  <li
    class="af-nav__item af-nav__item--haschild "
    role="none"
  >
    <span
      aria-expanded="false"
      class="af-nav__link"
      role="menuitem"
      tabindex="-1"
    >
      test
    </span>
    <i
      class="glyphicon glyphicon-arrow-xs-down"
    />
    <ul
      aria-label=""
      class="af-nav__list af-nav__list--children"
      role="menu"
    >
      <li
        class="af-nav__item "
        role="none"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Accueil"
          class="af-nav__link"
          href="/accueil"
          role="menuitem"
          tabindex="-1"
        >
          Accueil
        </a>
      </li>
      <li
        class="af-nav__item "
        role="none"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Membres"
          class="af-nav__link"
          href="/members"
          role="menuitem"
          tabindex="-1"
        >
          Membres
        </a>
      </li>
      <li
        class="af-nav__item "
        role="none"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Slash DS"
          class="af-nav__link"
          href="/slash"
          role="menuitem"
          tabindex="-1"
        >
          Slash DS
        </a>
      </li>
    </ul>
  </li>
</DocumentFragment>
`;

exports[`<MenuItemEnhanced/> Render <Menu/> with 1`] = `
<DocumentFragment>
  <li
    class="af-nav__item "
    role="none"
  >
    <a
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="test"
      class="af-nav__link"
      href="/demos/path"
      role="menuitem"
      tabindex="-1"
    >
      test
    </a>
  </li>
</DocumentFragment>
`;
